### PR TITLE
user delivery layer

### LINKED ISSUE #78: user delivery layer 구현

### PR DESCRIPTION
**Pull request by BLoHny**
# summary
- user usecase router add
# changes
- user usecase router add
- user input fix
- user test code add
- enum validator add

close #78

### CODE REVIEW

File: internal/app/user/delivery/user_router.go [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+	engine.GET("/rank", r.getRanking)
+	engine.GET("/users", r.search)
+}
+
+func (r *UserRouter) getUserDetail(c *gin.Context) {
+	id := c.Param("userID")
+
+	userID, err := strconv.ParseUint(id, 10, 64)
+	if err != nil {
+		c.Error(err)
```
Comment (onee-only) [inline internal/app/user/delivery/user_router.go]: 여기서 그냥 error를 넘기면 internal server error로 status가 넘어갑니다. status.NewError를 사용하는 게 어떨까요?

File: internal/app/user/delivery/user_router.go [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+		return
+	}
+
+	c.JSON(http.StatusOK, coins)
+}
+
+func (r *UserRouter) getMyDetail(c *gin.Context) {
+	res, err := r.userUseCase.GetMyDetail(c)
```
Comment (onee-only) [inline internal/app/user/delivery/user_router.go]: 여기만 output이 res라고 되어 있습니다. 다른 곳들 모두 res로 바꾸거나 여기도 detail같은 거로 바꾸는 게 좋을 것 같아요!

File: internal/app/user/delivery/user_router.go [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+	if err != nil {
+		c.Error(err)
```
Comment (onee-only) [inline internal/app/user/delivery/user_router.go:73]:
```suggestion
		c.Error(err)
		return
```

File: internal/app/user/delivery/user_router.go [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+	}
+
+	c.JSON(http.StatusOK, res)
+}
+
+func (r *UserRouter) getRanking(c *gin.Context) {
+	var rank Rank
+
+	if err := c.ShouldBindQuery(&rank); err != nil {
+		c.Error(err)
```
Comment (onee-only) [inline internal/app/user/delivery/user_router.go]: 여기도 status 써주세요

File: internal/app/user/delivery/user_router.go [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+
+	res, err := r.userUseCase.GetRanking(c, &input.RankQueryInput{
+		SortType: rank.SortType,
+		UserType: rank.UserType,
+		Grade:    rank.grade,
+		Class:    rank.class,
+		Name:     rank.name,
+	})
+	if err != nil {
+		c.Error(err)
```
Comment (onee-only) [inline internal/app/user/delivery/user_router.go:96]:
```suggestion
		c.Error(err)
		return
```

File: internal/app/user/delivery/user_router.go [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+	}
+
+	c.JSON(http.StatusOK, res)
+}
+
+func (r *UserRouter) search(c *gin.Context) {
+	var query Query
+
+	if err := c.ShouldBindQuery(&query); err != nil {
+		c.Error(err)
```
Comment (onee-only) [inline internal/app/user/delivery/user_router.go]: status 써주세요

File: internal/app/user/delivery/user_router.go [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+		return
+	}
+
+	res, err := r.userUseCase.Search(c, &input.SearchInput{
+		Query: query.Query,
+	})
+	if err != nil {
+		c.Error(err)
```
Comment (onee-only) [inline internal/app/user/delivery/user_router.go:115]:
```suggestion
		c.Error(err)
		return
```

File: internal/app/user/delivery/user_router.go [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+	SortType constant.SortType  `form:"sortType" binding:"required,enum"`
+	UserType *constant.UserType `form:"userType" binding:"enum"`
+	grade    *int               `form:"grade" binding:"number"`
+	class    *int               `form:"class" binding:"number"`
+	name     *string            `form:"name"`
+}
+
+type Query struct {
+	Query string
+}
```
Comment (onee-only) [inline internal/app/user/delivery/user_router.go]: request라는 게 명시되었으면 좋겠습니다! 그리고 파일 끝에 붙이는 것보단 따로 파일이나 패키지를 만들던가 이 파일 앞에 묶어 놓는 게 좋을 것 같습니다!

File: internal/app/user/usecase/get_ranking.go [hunk 9]
```diff
@@ -16,7 +16,7 @@ func (uc *userUseCase) GetRanking(ctx context.Context, input *input.RankQueryInp
 		err   error
 	)
 
-	if input.UserType == constant.TypeStudent {
+	if *input.UserType == constant.TypeStudent {
```
Comment (onee-only) [inline internal/app/user/usecase/get_ranking.go:19]: 이건 어떤 이유로 추가하셨나요?

File: internal/global/validator/validator.go [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+	"github.com/go-playground/validator/v10"
+	"github.com/lucky-pocket/luckyPocket-back/internal/domain/data/constant"
+)
+
+func Initialize() {
+	if v, ok := binding.Validator.Engine().(*validator.Validate); ok {
+		err := v.RegisterValidation("enum", enum)
+		if err != nil {
+			return
+		}
```
Comment (onee-only) [inline internal/global/validator/validator.go]:
```suggestion
		if err := v.RegisterValidation("enum", enum); err != nil {
			return
		}
```

File: internal/global/validator/validator.go [hunk 11]
```diff
@@ -0,0 +1,21 @@
+package validator
+
+import (
+	"github.com/gin-gonic/gin/binding"
+	"github.com/go-playground/validator/v10"
+	"github.com/lucky-pocket/luckyPocket-back/internal/domain/data/constant"
+)
+
+func Initialize() {
```
Comment (onee-only) [inline internal/global/validator/validator.go]:
```suggestion
func Initialize() error {
```
Initialize가 error를 반환하면 좋을 것 같습니다!

File: internal/app/user/usecase/get_ranking.go (hunk 9 above)
Comment (onee-only) [reply internal/app/user/usecase/get_ranking.go:19]: 들어온 input이 nil이면 panic이 날 것 같습니다. 이 부분은 따로 나중에 issue파서 해결 하는 게 어떨까요? 애초에 repository api가 nil인 것을 감안하지 않아서 따로 issue를 파는 게 맞을 것 같습니다.

File: internal/app/user/usecase/get_ranking.go [hunk 12]
```diff
@@ -16,7 +16,7 @@ func (uc *userUseCase) GetRanking(ctx context.Context, input *input.RankQueryInp
...
 		users, err = uc.UserRepository.RankStudents(ctx, input.SortType, input.Name, input.Class, input.Grade)
```
Comment (onee-only) [inline internal/app/user/usecase/get_ranking.go]: class랑 grade 자리가 바꿔졌습니다! 이것도 같이 해 주시면 감사하겠습니다!

File: test/unit/domain/user/router/get_rank_test.go [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+
+import (
+	"net/http"
+	"net/http/httptest"
+
+	"github.com/lucky-pocket/luckyPocket-back/internal/domain/data/output"
+	"github.com/stretchr/testify/mock"
+)
+
+func (s *UserRouterTestSuite) TestGetRank() {
```
Comment (onee-only) [inline test/unit/domain/user/router/get_rank_test.go:12]: 400이 뜨는 edge case도 test해 주셨으면 좋겠습니다!

File: test/unit/domain/user/router/get_user_detail_test.go [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+			statusCode: http.StatusNotFound,
+		},
+	}
+
+	for _, tc := range testcases {
+		s.Run(tc.desc, func() {
+			tc.on()
+
+			w := httptest.NewRecorder()
+			req, _ := http.NewRequest("GET", "/users/1", nil)
```
Comment (onee-only) [inline test/unit/domain/user/router/get_user_detail_test.go:39]: tc의 path가 사용되지 않는 것 같습니다! tc에 path를 지우거나 여기에 추가 해주세요!

File: test/unit/domain/user/router/suite_test.go [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+	mockUserUseCase *mocks.UserUseCase
+}
+
+func TestUserRouterSuite(t *testing.T) {
+	suite.Run(t, new(UserRouterTestSuite))
+}
+
+func (s *UserRouterTestSuite) SetupSuite() {
+	validator.Initialize()
+	s.mockUserUseCase = mocks.NewUserUseCase(s.T())
```
Comment (onee-only) [inline test/unit/domain/user/router/suite_test.go]:
```suggestion
	validator.Initialize()
	
	s.mockUserUseCase = mocks.NewUserUseCase(s.T())
```
이왕이면 띄어 놓는 게 좋을 것 같습니다!

File: internal/global/validator/validator.go (hunk 11 above)
Comment (onee-only) [reply internal/global/validator/validator.go]: 사실 Initialize가 *validator.Validate 타입을 인수로 받는 것도 괜찮을 것 같습니다. binding 호출은 main같은 바깥에서 하고요

File: internal/app/user/usecase/get_ranking.go (hunk 9 above)
Comment (BLoHny) [reply internal/app/user/usecase/get_ranking.go:19]: 추가하지 않으면 Mismatch 가 발생해서 추가해논 상태입니다

현재 UserType이 nil 허용이기때문에 로직자체를 수정하는게 나을것같습니다

Comment (onee-only) [reply internal/app/user/usecase/get_ranking.go:19]: 이슈 팠습니다

File: internal/app/user/delivery/user_router.go [hunk 16]
```diff
@@ -0,0 +1,120 @@
...
+	engine.GET("/rank", r.getRanking)
+	engine.GET("/users", r.search)
+}
+
+func (r *UserRouter) getUserDetail(c *gin.Context) {
+	id := c.Param("userID")
+
+	userID, err := strconv.ParseUint(id, 10, 64)
+	if err != nil {
+		status.NewError(http.StatusBadRequest, "not valid")
```
Comment (onee-only) [inline internal/app/user/delivery/user_router.go]:
```suggestion
		c.Error(status.NewError(http.StatusBadRequest, "not valid"))
```

File: internal/app/user/delivery/user_router.go [hunk 17]
```diff
@@ -0,0 +1,120 @@
...
+	}
+
+	c.JSON(http.StatusOK, userInfo)
+}
+
+func (r *UserRouter) getRanking(c *gin.Context) {
+	var rank RankRequest
+
+	if err := c.ShouldBindQuery(&rank); err != nil {
+		status.NewError(http.StatusBadRequest, "not valid param")
```
Comment (onee-only) [inline internal/app/user/delivery/user_router.go]:
```suggestion
		c.Error(status.NewError(http.StatusBadRequest, "not valid param"))
```

File: internal/app/user/delivery/user_router.go [hunk 18]
```diff
@@ -0,0 +1,120 @@
...
+		return
+	}
+
+	res, err := r.userUseCase.GetRanking(c, &input.RankQueryInput{
```
Comment (onee-only) [inline internal/app/user/delivery/user_router.go]: 여기에도 res 있습니다!

File: internal/app/user/delivery/user_router.go [hunk 19]
```diff
@@ -0,0 +1,120 @@
...
+	}
+
+	c.JSON(http.StatusOK, res)
+}
+
+func (r *UserRouter) search(c *gin.Context) {
+	var query QueryRequest
+
+	if err := c.ShouldBindQuery(&query); err != nil {
+		status.NewError(http.StatusBadRequest, "not valid param")
```
Comment (onee-only) [inline internal/app/user/delivery/user_router.go]:
```suggestion
		c.Error(status.NewError(http.StatusBadRequest, "not valid param"))
```

File: internal/app/user/delivery/user_router.go (hunk 18 above)
Comment (BLoHny) [reply internal/app/user/delivery/user_router.go]: [5fc9e57c120aefc559b4edbe5b96a42313f77bb1
](https://github.com/lucky-pocket/luckyPocket-back/pull/92/commits/5fc9e57c120aefc559b4edbe5b96a42313f77bb1)

**Comment by onee-only** [comment]
전체적으로 input에 c.Request.Context()를 넣어주세요!

File: internal/global/validator/validator.go (hunk 11 above)
Comment (BLoHny) [reply internal/global/validator/validator.go]: https://github.com/lucky-pocket/luckyPocket-back/pull/92/commits/7c631e272273b58eab70b7d99a2c7ec89018c677

Comment (BLoHny) [reply internal/global/validator/validator.go]: https://github.com/lucky-pocket/luckyPocket-back/pull/92/commits/82aafb981b33f44c6e2ad6c251202e5eb968c6eb

**Review by onee-only — APPROVED** [review]
